pkg/controller/federation/clusterdeployment_federation_controller.go

```diff
@@ -159,60 +161,71 @@ func (r *ReconcileClusterDeploymentFederation) Reconcile(request reconcile.Reque
 
 	// If already federated, skip
 	if cd.Status.Federated {
-		cdLog.Debug("cluster already federated, nothing to do")
+		// If already federated, ensure federated cluster is in sync
+		// with the cluster deployment
+		err = r.updateFederatedCluster(cd, cdLog)
+		if err != nil {
+			cdLog.WithError(err).Error("cannot update federated cluster")
+			return reconcile.Result{}, err
+		}
 		return reconcile.Result{}, nil
 	}
 
 	cdLog.Info("reconciling cluster deployment for federation")
 
+	if cd.Status.FederatedClusterRef == nil || len(cd.Status.FederatedClusterRef.Name) == 0 {
+		cdLog.Debugf("setting federated cluster name reference")
+		return reconcile.Result{}, r.setFederatedClusterRef(cd, cdLog)
+	}
```

Comment:
Looks like we generate and stamp a cluster ref before we even try to federate, so even if we fail it will re-try with the previous ref? Good thinking.

Reply:
Yup, otherwise retrying join would make no sense.